### PR TITLE
Don't allocate the sticky regex source when the resulting regexp is already cached

### PR DESCRIPTION
**Pull request by pygy**
As discussed previously in #152 and #153

### CODE REVIEW

File: src/xregexp.js [hunk 1]
```diff
@@ -839,7 +839,7 @@ XRegExp.exec = function(str, regex, pos, sticky) {
         regex[REGEX_DATA][cacheKey] = copyRegex(regex, {
             addG: true,
             addY: addY,
-            source: sourceY,
+            source: fakeY? regex.source + '|()' : void 0,
```
Comment (slevithan) [inline src/xregexp.js]: Nits: Space before `?`, and please just use `undefined` rather than `void 0`.

**Comment by pygy** [comment]
Thinking more about this, I'm not sure what to do for regexps that are both global and sticky. It makes little sense to set both flags, but it is legal AFAIK... Scaling down.

File: src/xregexp.js (hunk 1 above)
Comment (pygy) [inline src/xregexp.js]: Sure, updating

**Comment by slevithan** [comment]
To be clear, the "fast sticky polyfill" rev is just avoiding an unnecessary string concat in some cases, and otherwise not changing behavior at all, correct?

**Comment by pygy** [comment]
Correct. I've removed the flags thing, I'll make another PR once the `gy` situation is more clear.

**Comment by slevithan** [comment]
> I've removed the lastIndex thing, I'll make another PR once the `gy` situation is more clear.

Thanks--perfect! You removing it from this PR made me more confident to merge the other change. The `gy` situation was still a bit fuzzy to me even though you included some great tests (thank you!), so I was hoping to come back to that to examine more fully. BTW, the reason XRegExp isn't yet fully tackling `lastIndex` setting for `/y` is because most of the related code was written before the existence of ES6 `/y`. It was based on `/y` in old versions of Firefox. I appreciate you tackling this somewhat confusing area to bring XRegExp up to spec with ES6 in handling of `/g`, `/y`, and `/gy`.

**Comment by slevithan** [comment]
@pygy BTW, if you submit robust `/y` + `lastIndex` handling, I'd be happy to publish a v3.1.2 that gets your `/y` speed and correctness updates out there sooner.

**Comment by pygy** [comment]
Ok, According to both the [latest draft](https://tc39.github.io/ecma262/#sec-regexpbuiltinexec) and the [ES6](http://www.ecma-international.org/ecma-262/6.0/#sec-regexpbuiltinexec) spec, `y` takes precedence over `g`. I'll look into it tomorrow evening.

**Pull request merged**